viewing_party/party.py

```diff
@@ -1,23 +1,241 @@
+        }
+    '''
+
+    lod_watched = user_data["watched"]
+    total_rating = 0
+    
+    if lod_watched == []:
+        return 0.0
+    for movie in lod_watched:
+        total_rating += movie["rating"]
+    
+    avg_rating = total_rating/len(lod_watched)
+    return avg_rating
         
-# -----------------------------------------
-# ------------- WAVE 4 --------------------
-# -----------------------------------------
+
+def get_most_watched_genre(user_data):
+    '''
+    return the genre that is the most frequently watched
+    If the value of "watched" is an empty list, get_most_watched_genre should return None.
+    user_data = {
+        'watched': [{"title": "", "genre": "", "rating": <float>}, {...}, {...}, {...}, {...}, {...}]
+        }
+    '''
+
+    lod_watched = user_data["watched"]
+    genre_count = {}
+
+
+    if lod_watched == []:
+        return None
+    for movie in lod_watched:
+        genre = movie["genre"] # assigning value of genre to variable "genre"
+        count = genre_count.get(genre, 0)
+        genre_count[genre] = count + 1
+
+    # Determine the most frequent
+    max_count = 0
+    frequent_genre = None
+
+    for key, value in genre_count.items():
+        if value > max_count:
+            max_count = value
+            frequent_genre = key
+    return frequent_genre
 
 # -----------------------------------------
-# ------------- WAVE 5 --------------------
+# ------------- WAVE 3 --------------------
 # -----------------------------------------
 
+def get_unique_watched(user_data):
+    '''
+    Return a list of dictionaries, that represents a list of movies that:
+    -the user has watched, but none of their friends have watched. (user movies) - (friends movies)
+    user_data = {
+        'watched': [{...}, {...}, {...}, {...}, {...}, {...}], 
+        friends': [{'watched': [..{},{},{}..], {'watched': [..{},{},{}..]
+        }
+    '''
+    titles_friends_watched = set()
+
+    for friend_movie_dict in user_data["friends"]:
+        for movie in friend_movie_dict["watched"]:
+            titles_friends_watched.add(movie["title"])
+    
+    # unique list of movies user's watched but not their friends
+    list_unique_movies = []
+    for movie in user_data["watched"]:
+        if movie["title"] not in titles_friends_watched:
+            list_unique_movies.append(movie)
+            
+    return list_unique_movies
+
+def get_friends_unique_watched(user_data):
+    '''
+    Return a list of dictionaries, that represents a list of movies that:
+    - aleast one of the user's friends have watched, but the user has not watched aka (friends movies) - (user movies)
+    user_data = {
+        'watched': [{...}, {...}, {...}, {...}, {...}, {...}], 
+        friends': [{'watched': [..{},{},{}..], {'watched': [..{},{},{}..]
+        }
+    '''
+
+    list_titles_watched = set()
+    for movie in user_data["watched"]:
+        list_titles_watched.add(movie["title"])

```

Comment:
In this function, you build up a list of all the movies a user has watched and then you iterate through all the movies a user's friends have watched to create `list_friends_unique_movie`.

What if you flipped the logic so that you first get all the movies the friends have watched (like you did in `get_unique_watched`) and then you can check that the friends' movies list are not in the user's movies list?

If you invert the logic, then you could use reuse the hypothetical helper function that gets friends' movies here too and keep this logic to a single responsibility. 

Like this:
```python
def get_friends_unique_watched(user_data):
    friends_watched_movies = get_friends_movies()

    list_friends_unique_movie = []

    for movie in friends_watched_movies:
         if movie not in user_data["watched"] and movie not in list_friends_unique_movie:
                list_friends_unique_movie.append(movie)

    return list_friends_unique_movie
```